### PR TITLE
Importation classes with imported name and alias

### PR DESCRIPTION
**Pull request by jayvdb**
In order to solve many corner cases related to imports,
more information is needed about each import.

This change creates two new classes:
- SubmoduleImportation
- ImportationFrom

And adds an optional parameter full_name to the super class
Importation.

Functionally, this change only improves existing error messages
to report the full imported name where previously an error
would include only the import alias.

### CODE REVIEW

**Comment by jayvdb** [comment]
`source_statement` exists as I would like to include that in a better `__repr__`

Note that I have a functionally equivalent changeset which does all the work within the existing `Importation` class: https://github.com/jayvdb/pyflakes/commit/e250ea6085a15485889115b8524d616254e127aa . IMO that isnt as neat and tidy; the special case for `_submodule_import` adds complexity throughout.

**Comment by jayvdb** [comment]
@bitglue , could you please review this.  I have pending changes that depend on this.

**Comment by bitglue** [comment]
Looks good on a brief read. I'll take a deeper look tonight.

**Comment by bitglue** [comment]
This looks good. Can you please rebase onto master?

File: pyflakes/checker.py [hunk 1]
```diff
@@ -136,16 +136,83 @@ class Importation(Definition):
...
+            return self.fullName
+
+
+class SubmoduleImportation(Importation):
+
+    def __init__(self, name, source):
+        # A dot should only appear in the name when it is a submodule import
+        # without an 'as' clause, which is a special type of import where the
+        # root module is implicitly imported, and the submodules are also
+        # imported by Python does not restrict which attributes of the root
```
Comment (jayvdb) [inline pyflakes/checker.py]: ah, just noticed : 'by' should be 'but' or 'however'

File: pyflakes/checker.py [hunk 2]
```diff
@@ -136,16 +136,83 @@ class Importation(Definition):
...
+        """Generate a source statement equivalent to the import."""
+        if self._has_alias():
+            return 'import %s as %s' % (self.fullName, self.name)
+        else:
+            return 'import %s' % self.fullName
+
+    def __str__(self):
+        """Return import full name with alias."""
+        if self._has_alias():
+            return self.fullName + ' as ' + self.name
```
Comment (jayvdb) [inline pyflakes/checker.py]: would we prefer `%s` syntax here instead of `+` ?  It does seem that %s syntax is preferred.

Comment (bitglue) [inline pyflakes/checker.py]: I don't have any particular preference.

**Comment by collinanderson** [comment]
This appears to break relative imports like: `from . import urls`

**Comment by jayvdb** [comment]
Yea, there were no tests covering them, and I forgot about those.  The fix #61 is waiting for review.

**Pull request merged**